examples/trading-human-loop/agent.py

```diff
@@ -0,0 +1,145 @@
+import asyncio
+import os
+from pathlib import Path
+
+from dotenv import load_dotenv
+
+from galadriel import AgentRuntime, LiteLLMModel
+from galadriel.agent import CodeAgent, ToolCallingAgent
+from galadriel.clients import Cron, ChatUIClient, DiscordClient, TerminalClient
+from galadriel.entities import Message, PushOnlyQueue
+from galadriel.memory.memory_store import MemoryStore
+from galadriel.wallets.solana_wallet import SolanaWallet
+from galadriel.tools.web3.market_data import coingecko, dexscreener
+from galadriel.tools.web3.onchain.solana import (
+    jupiter,
+    raydium,
+    native as solana_native,
+    spl_token,
+)
+
+TRADING_INTERVAL_SECONDS = 600
+
+PROMPT= """
+You are a helpful trading agent collaborating with a human trader in the Solana ecosystem.
+
+### Chat History:
+{{chat_history}}
+
+### Request:
+{{request}}
+
+Use the chat history to provide relevant insights. Your goal is to identify the best trading opportunities while ensuring human confirmation before executing any trade. This is crucial.
+
+**Guidelines:**
+- Present trade analysis and top 3 concrete investment options, always inclue a 4th option to not trade.
+- Make each option short (couple of sentences max) and add data to support the option if possible.
+- Confirm with the trader before executing any trade by checking the most recent chat history.
+- If confirmation is being given in the current request, proceed with execution; otherwise, ask again.
+- Maintain a chat-like conversation with the human trader, this is also important.
+- Chat history is important but it might be empty at the beginning of the conversation.
+- VERY IMPORTANT: if you have the final answer, wrap it in the appropriate manner using a code block function. call the final_answer tool with a string as argument.
+
+"""
+
+load_dotenv(dotenv_path=Path(".") / ".env", override=True)
+load_dotenv(dotenv_path=Path(".") / ".agents.env", override=True)
+
+model = LiteLLMModel(
+    model_id="openai/gpt-4o",
+    api_key=os.getenv("OPENAI_API_KEY"),
+)
+
+solana_wallet = SolanaWallet(key_path=os.getenv("SOLANA_KEY_PATH"))
+
+tools = [
+    coingecko.GetMarketDataPerCategoriesTool(),
```

Comment:
btw I saw in a video specialising in agents that the optimum number of tools for GPT-o4 seems to be 4-6.